lib/widgets/actions.dart

```diff
@@ -16,101 +16,171 @@ import '../model/narrow.dart';
+  try {
+    final store = PerAccountStoreWidget.of(context);
+    final connection = store.connection;
+    final useLegacy = connection.zulipFeatureLevel! < 155; // TODO(server-6)
+    if (useLegacy) {
+      await _legacyMarkNarrowAsRead(context, narrow);
+      return;
+    }
 
-  while (true) {
-    final result = await updateMessageFlagsForNarrow(connection,
-      anchor: anchor,
+    // Compare web's `mark_all_as_read` in web/src/unread_ops.js
+    // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    final didPass = await updateMessageFlagsStartingFromAnchor(
+      context: context,
+      // Include `is:unread` in the narrow.  That has a database index, so
+      // this can be an important optimization in narrows with a lot of history.
+      // The server applies the same optimization within the (deprecated)
+      // specialized endpoints for marking messages as read; see
+      // `do_mark_stream_messages_as_read` in `zulip:zerver/actions/message_flags.py`.
+      apiNarrow: narrow.apiEncode()..add(ApiNarrowIsUnread()),
+      // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
+      // will be the oldest non-muted unread message, which would
+      // result in muted unreads older than the first unread not
+      // being processed.
+      startingAnchor: AnchorCode.oldest,
       // [AnchorCode.oldest] is an anchor ID lower than any valid
       // message ID; and follow-up requests will have already
       // processed the anchor ID, so we just want this to be
       // unconditionally false.
       includeAnchor: false,
-      // There is an upper limit of 5000 messages per batch
-      // (numBefore + numAfter <= 5000) enforced on the server.
-      // See `update_message_flags_in_narrow` in zerver/views/message_flags.py .
-      // zulip-mobile uses `numAfter` of 5000, but web uses 1000
-      // for more responsive feedback. See zulip@f0d87fcf6.
-      numBefore: 0,
-      numAfter: 1000,
-      narrow: apiNarrow,
       op: UpdateMessageFlagsOp.add,
-      flag: MessageFlag.read);
-    if (!context.mounted) {
-      scaffoldMessenger.clearSnackBars();
-      return;
+      flag: MessageFlag.read,
+      onCompletedMessage: zulipLocalizations.markAsReadComplete,
+      progressMessage: zulipLocalizations.markAsReadInProgress,
+      onFailedTitle: zulipLocalizations.errorMarkAsReadFailedTitle);
+
+    if (!didPass || !context.mounted) return;
+    if (narrow is CombinedFeedNarrow && !useLegacy) {
+      PerAccountStoreWidget.of(context).unreads.handleAllMessagesReadSuccess();
     }
-    responseCount++;
-    updatedCount += result.updatedCount;
+  } catch (e) {
+    if (!context.mounted) return;
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    await showErrorDialog(context: context,
+      title: zulipLocalizations.errorMarkAsReadFailedTitle,
```

Comment:
Hmm — it doesn't feel right that we have two of these generic try/catch blocks nested one inside the other, and both reporting the same error. That feels like a sign that one of them is too broad.

This is what caused me to notice the point mentioned above, about the handleAllMessagesReadSuccess call getting moved inside the try/catch.

Then after that's fixed, I think the remaining thing this try/catch is covering is basically the _legacyMarkNarrowAsRead call. (There's no way updateMessageFlagsStartingFromAnchor can throw, short of a bug in its `catch` block; and there's little else here, and none of it can throw except through some surprising bug in our code.) So the remaining thing to do to fix this redundancy is to move this second try/catch to be inside that function.